sdk/include/opentelemetry/sdk/trace/simple_processor.h

```diff
@@ -48,11 +57,29 @@ class SimpleSpanProcessor : public SpanProcessor
 
   void Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override
   {
+    WaitLock();
     exporter_->Shutdown(timeout);
+    ClearLock();
   }
 
 private:
   std::unique_ptr<SpanExporter> exporter_;
+  std::atomic_flag lock_;
```

Comment:
Put `lock_` into RAII class and make unblock to be exception safe?

Reply:
Good point. Introducing something similar to `std::lock_guard` would be great. We could also use this [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/provider.h#L26).

Reply:
Sounds reasonable.  

For reference, where do you want utilities like this?  `nostd`? `api` (since we'll be using it for the trace provider as well)?

Reply:
Let's put it into the `api/include/opentelemetry/trace/common` folder and namespace. In `nostd` we only put stuff that mimics the STL stuff and for which we could possible use standardized STL classes as a drop-in replacement (e. g. `shared_ptr`, `variant`, `unique_ptr`, `span`).

As far as I know, the STL doesn't offer a `lock_guard` for atomic flags. So this should go into `common`.

Reply:
The [meter provider](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/metrics/provider.h#L26) is another place where we should use the common implementation.

Reply:
Another possibility for a solution: you could provide a mutex implementation based on an `atomic_flag` (with `Lock` and `Unlock` operations, like shown [here](https://stackoverflow.com/questions/43546213/stdatomic-flag-and-stdlock-guard)).

Then we could use `std::lock_guard` around that custom mutex implementation.

No the most efficient, but all the places where we use it are not on code paths that are critical for performance.

Reply:
I've added a commit which does the following:

1. Added a `SpinLockMutex` class in api/common.
2. Wired the Mutex into existing spin-lock locations (`*Providers`).
3. Updated the docs around thread-safety for `Exporter` and `SpanProcessor`
4. Modified the `SimpleSpanProcessor`.



## Side Note:
The single-threaded portion of the spec may be inconsistently enforced. I looked at Java to see what set up they are using for locking, and found this: https://github.com/open-telemetry/opentelemetry-java/blob/71ba8e1b87bfd16f2ecd21643bf55b88d1ecfcaa/sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/SimpleSpanProcessor.java

**TL;DR:**
- The result of Export is a 'promise'
- There appears to be no explicit locking (outside of registering callbacks) on the SpanProcessor
- I'm not sure this translates (well) to C++.